File: scala/src/main/scala/ComplexNumber.scala
class ComplexNumber(val real:Double, val imaginary:Double) {
  def +(that: ComplexNumber): ComplexNumber = new ComplexNumber(
    this.real + that.real,
    this.imaginary + that.imaginary)
  
  def -(that: ComplexNumber): ComplexNumber = new ComplexNumber(
    this.real - that.real,
    this.imaginary - that.imaginary)
  
  def *(that: ComplexNumber): ComplexNumber = new ComplexNumber(
    this.real * that.real - this.imaginary * that.imaginary,
    this.imaginary * that.real + this.real * that.imaginary,
  )
  
  def /(that: ComplexNumber): ComplexNumber = {
    val size = that.abs2
    new ComplexNumber(
      (this.real * that.real + this.imaginary * that.imaginary) / size,
      (this.imaginary * that.real - this.real * that.imaginary) / size,
    )
  }
  
  def conjugate: ComplexNumber = new ComplexNumber(this.real, -this.imaginary)
  
  def abs2: Double = this.real * this.real + this.imaginary * this.imaginary

  def abs: Double = Math.sqrt(this.abs2)
}

object ComplexNumber {
    def apply(real:Double = 0, imaginary:Double = 0): ComplexNumber = {
        new ComplexNumber(real, imaginary)
    }

    def exp(that: ComplexNumber): ComplexNumber = {
      val size = Math.exp(that.real)
      new ComplexNumber(
        size * Math.cos(that.imaginary), 
        size * Math.sin(that.imaginary)
      )
    }
}

File: java/src/main/java/ComplexNumber.java
class ComplexNumber {
    private final double real;
    private final double imaginary;

    ComplexNumber(final double real, final double imaginary) {
        this.real = real;
        this.imaginary = imaginary;
    }

    double getReal() {
        return real;
    }

    double getImaginary() {
        return imaginary;
    }

    double abs2() {
        return real * real + imaginary * imaginary;
    }

    double abs() {
        return Math.sqrt(abs2());
    }

    ComplexNumber add(final ComplexNumber other) {
        return new ComplexNumber(real + other.real, imaginary + other.imaginary);
    }

    ComplexNumber subtract(final ComplexNumber other) {
        return new ComplexNumber(real - other.real, imaginary - other.imaginary);
    }

    ComplexNumber multiply(final ComplexNumber other) {
        return new ComplexNumber(
                (real * other.real - imaginary * other.imaginary),
                (imaginary * other.real + real * other.imaginary)
        );
    }

    ComplexNumber multiply(final double factor) {
        return new ComplexNumber(real * factor, imaginary * factor);
    }

    ComplexNumber divide(final ComplexNumber other) {
        final double divisor = other.abs2();
        return new ComplexNumber(
                (real * other.real + imaginary * other.imaginary)/divisor,
                (imaginary * other.real - real * other.imaginary)/divisor
        );
    }

    ComplexNumber divide(final double divisor) {
        return new ComplexNumber(real / divisor, imaginary / divisor);
    }

    ComplexNumber conjugate() {
        return new ComplexNumber(real, -imaginary);
    }

    ComplexNumber exponentialOf() {
        final double expReal = Math.exp(real);
        return new ComplexNumber(
                expReal * Math.cos(imaginary),
                expReal * Math.sin(imaginary)
        );
    }

}